src/fdb5/remote/client/RemoteCatalogue.cc

```diff
@@ -0,0 +1,209 @@
+/*
+ * (C) Copyright 1996- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation nor
+ * does it submit to any jurisdiction.
+ */
+
+#include "eckit/config/Resource.h"
+#include "eckit/log/Log.h"
+#include "eckit/serialisation/MemoryStream.h"
+
+#include "fdb5/LibFdb5.h"
+#include "fdb5/remote/client/RemoteCatalogue.h"
+
+#include <unordered_map>
+
+using namespace eckit;
+namespace fdb5::remote {
+
+RemoteCatalogue::RemoteCatalogue(const Key& key, const Config& config):
+    CatalogueImpl(key, ControlIdentifiers(), config), // xxx what are control identifiers? Setting empty here...
+    Client(eckit::net::Endpoint(config.getString("host"), config.getInt("port")), ""),
+    config_(config), schema_(nullptr), numLocations_(0) {
+
+    loadSchema();
+}
+
+RemoteCatalogue::RemoteCatalogue(const eckit::URI& uri, const Config& config):
+    Client(eckit::net::Endpoint(config.getString("host"), config.getInt("port")), ""), config_(config), schema_(nullptr), numLocations_(0)
+    {
+        NOTIMP;
+    }
+
+
+void RemoteCatalogue::sendArchiveData(uint32_t id, const Key& key, std::unique_ptr<FieldLocation> fieldLocation)
+{
+    ASSERT(!dbKey_.empty());
+    ASSERT(!currentIndexKey_.empty());
+    ASSERT(!key.empty());
+    ASSERT(fieldLocation);
+
+    Buffer keyBuffer(4096);
+    MemoryStream keyStream(keyBuffer);
+    keyStream << currentIndexKey_;
+    keyStream << key;
+
+    Buffer locBuffer(4096);
+    MemoryStream locStream(locBuffer);
+    locStream << *fieldLocation;
+
+    std::vector<std::pair<const void*, uint32_t>> payloads;
+    payloads.push_back(std::pair<const void*, uint32_t>{keyBuffer, keyStream.position()});
+    payloads.push_back(std::pair<const void*, uint32_t>{locBuffer, locStream.position()});
+
+    dataWrite(Message::Blob, id, payloads);
```

Comment:
There is serialization and deserialization code in many places and I think this needs to be abstracted away into the read/write methods.

With the current implementation I see problems if we ever want to evolve the protocol, either by changing the serialization format or by introducing new / deprecating old fields.

I would like to see something along the lines of:

```cpp
struct MessageBlob {
  // Contains everything the message entails
};

auto msg = createMessage(...) // Or use a builder
write(msg) // This takes care of buffer handling & serialization
```